rust/src/operations/optimize.rs

```diff
@@ -262,101 +333,153 @@ pub struct MergePlan {
-                        file_meta.location.as_ref(),
-                        partition_values,
-                        size,
-                    )?);
-
-                    metrics.num_files_removed += 1;
-                    metrics.files_removed.total_files += 1;
-                    metrics.files_removed.total_size += file_meta.size as i64;
-                    metrics.files_removed.max = std::cmp::max(metrics.files_removed.max, size);
-                    metrics.files_removed.min = std::cmp::min(metrics.files_removed.min, size);
-                }
 
-                // Save the file to storage and create corresponding add and remove actions. Do not commit yet.
-                let add_actions = writer.close().await?;
-                if add_actions.len() != 1 {
-                    // Ensure we don't deviate from the merge plan which may result in idempotency being violated
-                    return Err(DeltaTableError::Generic(
-                        "Expected writer to return only one add action".to_owned(),
-                    ));
-                }
-                for mut add in add_actions {
-                    add.data_change = false;
-                    let size = add.size;
-
-                    metrics.num_files_added += 1;
-                    metrics.files_added.total_files += 1;
-                    metrics.files_added.total_size += size;
-                    metrics.files_added.max = std::cmp::max(metrics.files_added.max, size);
-                    metrics.files_added.min = std::cmp::min(metrics.files_added.min, size);
-                    actions.push(action::Action::add(add));
-                }
+        // Need to move metrics and operations out of self, so we can use self in the stream
+        let mut metrics = std::mem::take(&mut self.metrics);
+
+        let operations = std::mem::take(&mut self.operations);
+
+        match operations {
+            OptimizeOperations::Compact(bins) => {
+                futures::stream::iter(bins)
+                    .flat_map(|(partition, bins)| {
+                        futures::stream::iter(bins).map(move |bin| (partition.clone(), bin))
+                    })
+                    .map(|(partition, files)| async {
+                        let rewrite_result = tokio::task::spawn(Self::rewrite_files(
+                            self.task_parameters.clone(),
+                            partition,
+                            files,
+                            object_store.clone(),
+                        ))
+                        .await;
+                        match rewrite_result {
+                            Ok(Ok(data)) => Ok(data),
+                            Ok(Err(e)) => Err(e),
+                            Err(e) => Err(DeltaTableError::GenericError {
+                                source: Box::new(e),
+                            }),
+                        }
+                    })
+                    .buffer_unordered(self.max_concurrent_tasks)
```

Comment:
I'm finding this is a good pattern for parallel processing a stream:

1. Map tasks into `tokio::task::spawn()` (which allows them to be sent to other threads)
2. Then follow with `buffer_unordered` to control the number of tasks being spawned at a time.